cmd/ceremony/key.go

```diff
@@ -0,0 +1,158 @@
+package main
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+
+	"github.com/letsencrypt/boulder/pkcs11helpers"
+	"github.com/miekg/pkcs11"
+)
+
+type generateArgs struct {
+	mechanism    []*pkcs11.Mechanism
+	privateAttrs []*pkcs11.Attribute
+	publicAttrs  []*pkcs11.Attribute
+}
+
+func getRandomBytes(ctx pkcs11helpers.PKCtx, session pkcs11.SessionHandle) ([]byte, error) {
+	r, err := ctx.GenerateRandom(session, 4)
+	if err != nil {
+		return nil, err
+	}
+	return r, nil
+}
+
+const (
+	rsaModLen = 2048
+	rsaExp    = 65537
+)
+
+type keyInfo struct {
+	key interface{}
+	der []byte
+	id  []byte
+}
+
+func generateKey(ctx pkcs11helpers.PKCtx, session pkcs11.SessionHandle, config ceremonyConfig) (*keyInfo, error) {
+	var pubKey interface{}
+	var keyID []byte
+	var err error
+	switch config.KeyType {
+	case "RSA":
+		pubKey, keyID, err = rsaGenerate(ctx, session, config.KeyLabel, rsaModLen, rsaExp)
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate RSA key pair: %s", err)
+		}
+	case "ECDSA":
+		pubKey, keyID, err = ecGenerate(ctx, session, config.KeyLabel, config.ECDSACurve)
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate ECDSA key pair: %s", err)
+		}
+	}
+
+	der, err := x509.MarshalPKIXPublicKey(pubKey)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to marshal public key: %s", err)
+	}
+
+	pemBytes := pem.EncodeToMemory(&pem.Block{Type: "PUBLIC KEY", Bytes: der})
+	log.Printf("Public key PEM:\n%s\n", pemBytes)
+	if err := ioutil.WriteFile(config.PublicKeyPath, pemBytes, 0644); err != nil {
+		return nil, fmt.Errorf("Failed to write public key to %q: %s", config.PublicKeyPath, err)
+	}
+	log.Printf("Public key written to %q\n", config.PublicKeyPath)
+	return &keyInfo{key: pubKey, der: der, id: keyID}, nil
+}
+
+// findObject looks up a PKCS#11 object handle based on the provided template.
+// In the case where zero or more than one objects are found to match the
+// template an error is returned.
+func findObject(ctx pkcs11helpers.PKCtx, session pkcs11.SessionHandle, tmpl []*pkcs11.Attribute) (pkcs11.ObjectHandle, error) {
```

Comment:
I think findObject and getKey should be moved out into their own pkcs11.go file.

Reply:
Part of the CRL work moved this into `pkcs11helpers`, might make sense to just do that work here instead of putting it off until that changeset.

Reply:
Ping on this feedback (moving findObject and getKey to their own file).

Reply:
I've moved `findObject` to the `pkcs11helpers` package. I think it makes sense to keep `getKey` in `cmd/ceremony` though since it is quite specific to this application.

That said I think it probably makes sense to a) change the name to `newSigner` and b) move it to `cert.go`, since it is really part of the `x509Signer` code than the key generation stuff.